cudaextender/tests/Test_CudaextenderEnd2End.cu

```diff
@@ -0,0 +1,126 @@
+    test_cases.push_back(test1);
+
+    return test_cases;
+}
+
+// Testing correctness of UngappedXDrop extension End2End with default scoring matrix
+class TestCudaextenderEnd2End : public ::testing::TestWithParam<End2EndTestParam>
+{
+public:
+    void SetUp()
+    {
+        param_ = GetParam();
+        // Define Scoring Matrix
+        const int32_t score_matrix[NUC2] = {91, -114, -31, -123, -1000, -1000, -100, -9100,
+                                            -114, 100, -125, -31, -1000, -1000, -100, -9100,
+                                            -31, -125, 100, -114, -1000, -1000, -100, -9100,
+                                            -123, -31, -114, 91, -1000, -1000, -100, -9100,
+                                            -1000, -1000, -1000, -1000, -1000, -1000, -1000, -9100,
+                                            -1000, -1000, -1000, -1000, -1000, -1000, -1000, -9100,
+                                            -100, -100, -100, -100, -1000, -1000, -100, -9100,
+                                            -9100, -9100, -9100, -9100, -9100, -9100, -9100, -9100};
+        const int32_t xdrop_threshold    = 910;
+        const int32_t device_id          = 0;
+        const bool no_entropy            = false;
+        const std::size_t max_gpu_memory = cudautils::find_largest_contiguous_device_memory_section();
+        allocator_                       = create_default_device_allocator(max_gpu_memory);
+        stream_                          = make_cuda_stream();
+        ungapped_extender_               = create_extender(score_matrix, NUC2, xdrop_threshold, no_entropy, stream_.get(), device_id, allocator_);
+    }
+
+    void TearDown()
+    {
+        ungapped_extender_.reset();
+    }
+
+    void run_test()
+    {
+        std::unique_ptr<io::FastaParser> fasta_parser_target = io::create_kseq_fasta_parser(param_.target_file, 0, false);
+        const std::string target_sequence                    = fasta_parser_target->get_sequence_by_id(0).seq;
+        std::unique_ptr<io::FastaParser> fasta_parser_query  = io::create_kseq_fasta_parser(param_.query_file, 0, false);
+        const std::string query_sequence                     = fasta_parser_query->get_sequence_by_id(0).seq;
+        std::vector<SeedPair> h_seed_pairs;
+        parse_seed_pairs(param_.seed_pairs_file, h_seed_pairs);
+        // Allocate pinned memory for query and target strings
+        pinned_host_vector<int8_t> h_encoded_target(target_sequence.length());
+        pinned_host_vector<int8_t> h_encoded_query(target_sequence.length());
+        encode_sequence(h_encoded_target.data(), target_sequence.c_str(), target_sequence.length());
+        encode_sequence(h_encoded_query.data(), query_sequence.c_str(), query_sequence.length());
+        ungapped_extender_->extend_async(h_encoded_query.data(), h_encoded_query.size(), h_encoded_target.data(), h_encoded_target.size(), param_.score_threshold, h_seed_pairs);
```

Comment:
Since the API accepts `int32_t`, please use `get_size<int32_t>(container)` instead of `container.size()`.

Reply:
Done.